File: LeetCode/Java/547-union-find-朋友圈.java
/*
 * @lc app=leetcode.cn id=547 lang=java
 *
 * [547] 朋友圈
 */

// @lc code=start
class Solution {

    class Union {
        private int count;
        private int[] parent;

        public Union(int n) {
            count = n;
            parent = new int[n];
            for (int i = 0; i < n; i++)
                parent[i] = i;
        }

        private int find(int p) {
            int root = p;
            while (parent[root] != root)
                root = parent[root];
            while (parent[p] != p) {
                int t = p;
                p = parent[p];
                parent[t] = root;
            }
            return root;
        }

        public void union(int p, int q) {
            int rootP = find(p);
            int rootQ = find(q);
            if (rootP == rootQ)
                return;
            parent[rootP] = rootQ;
            --count;
        }

        public int count() {
            return count;
        }
    }

    public int findCircleNum(int[][] M) {
        int n = M.length;
        var union = new Union(n);
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++)
                if (M[i][j] == 1)
                    union.union(i, j);
        return union.count();
    }
}
// @lc code=end

File: LeetCode/Go/547-union-find-朋友圈.go
/*
 * @lc app=leetcode.cn id=547 lang=golang
 *
 * [547] 朋友圈
 */
package leetcode

// @lc code=start
func findCircleNum(M [][]int) int {
	if len(M) == 0 {
		return 0
	}
	n := len(M)
	count := n
	parent := []int{}
	for i := 0; i < n; i++ {
		parent = append(parent, i)
	}

	var find func(int) int
	find = func(p int) int {
		if parent[p] != p {
			parent[p] = find(parent[p])
		}
		return parent[p]
	}

	union := func(p, q int) {
		rp, rq := find(p), find(q)
		if rp != rq {
			parent[rp] = rq
			count--
		}
	}

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if M[i][j] == 1 {
				union(i, j)
			}
		}
	}
	return count
}

// @lc code=end
